velox/functions/sparksql/String.h

```diff
@@ -455,4 +455,139 @@ struct SubstrFunction {
   }
 };
 
+struct OverlayFunctionBase {
+  template <bool isAscii, bool isVarchar>
+  FOLLY_ALWAYS_INLINE void doCall(
+      exec::StringWriter<false>& result,
+      const StringView& input,
+      const StringView& replace,
+      const int32_t pos,
+      const int32_t len) {
+    // Calculate and append first part.
+    auto pair = substring<isAscii, isVarchar>(input, 1, pos - 1);
+    if constexpr (isVarchar) {
+      auto byteRange = stringCore::getByteRange<isAscii>(
+          input.data(), pair.first + 1, pair.second);
+      result.append(StringView(
+          input.data() + byteRange.first, byteRange.second - byteRange.first));
+    } else {
+      result.append(StringView(input.data() + pair.first, pair.second));
+    }
+
+    // Append second part.
+    result.append(replace);
+
+    // Calculate and append last part.
+    int32_t length = 0;
+    if constexpr (isVarchar) {
+      length = len >= 0 ? len : stringImpl::length<isAscii>(replace);
+    } else {
+      length = len >= 0 ? len : replace.size();
+    }
+    int64_t start = (int64_t)pos + (int64_t)length;
+    pair = substring<isAscii, isVarchar>(input, start, INT32_MAX);
+    if constexpr (isVarchar) {
+      auto byteRange = stringCore::getByteRange<isAscii>(
+          input.data(), pair.first + 1, pair.second);
+      result.append(StringView(
+          input.data() + byteRange.first, byteRange.second - byteRange.first));
+    } else {
+      result.append(StringView(input.data() + pair.first, pair.second));
+    }
+  }
+
+  // Information regarding the pos calculation:
+  // Hive and SQL use one-based indexing for SUBSTR arguments but also accept
+  // zero and negative indices for start positions. If a start index i is
+  // greater than 0, it refers to element i-1 in the sequence. If a start index
+  // i is less than 0, it refers to the -ith element before the end of the
+  // sequence. If a start index i is 0, it refers to the first element. Return
+  // pair of first indices and length.
+  template <bool isAscii, bool isVarchar>
+  FOLLY_ALWAYS_INLINE std::pair<int32_t, int32_t>
+  substring(const StringView& input, const int64_t pos, const int64_t length) {
+    int64_t len = 0;
+    if constexpr (isVarchar) {
+      len = stringImpl::length<isAscii>(input);
+    } else {
+      len = input.size();
+    }
+    int64_t start = (pos > 0) ? pos - 1 : ((pos < 0) ? len + pos : 0);
+    int64_t end = 0;
+    if (start + length > INT32_MAX) {
```

Comment:
Where is INT32_MAX defined?

Reply:
It defined in stdint.h, I see other tests also use it.